File: proton-sdk/client/cs/src/Proton.Drive.Sdk/Nodes/Revision.cs
using Proton.Sdk;

namespace Proton.Drive.Sdk.Nodes;

public sealed record Revision
{
    public required RevisionUid Uid { get; init; }
    public required DateTime CreationTime { get; init; }
    public required long SizeOnCloudStorage { get; init; }
    public long? ClaimedSize { get; init; }
    public FileContentDigests ClaimedDigests { get; init; }
    public DateTime? ClaimedModificationTime { get; init; }
    public required IReadOnlyList<ThumbnailHeader> Thumbnails { get; init; }
    public required IReadOnlyList<AdditionalMetadataProperty>? AdditionalClaimedMetadata { get; init; }
    public Result<Author, SignatureVerificationError>? ContentAuthor { get; init; }
}

File: proton-sdk/incubating/client/kt/sdk/src/main/kotlin/me/proton/drive/sdk/extension/Revision.kt
package me.proton.drive.sdk.extension

import me.proton.drive.sdk.entity.FileRevision
import me.proton.drive.sdk.entity.RevisionUid
import proton.drive.sdk.ProtonDriveSdk
import proton.drive.sdk.claimedDigestsOrNull
import proton.drive.sdk.claimedModificationTimeOrNull
import proton.drive.sdk.contentAuthorOrNull

fun ProtonDriveSdk.FileRevision.toEntity() = FileRevision(
    uid = RevisionUid(uid),
    creationTime = creationTime.toInstant(),
    sizeOnCloudStorage = sizeOnCloudStorage,
    claimedSize = if (hasClaimedSize()) claimedSize else null,
    claimedDigests = claimedDigestsOrNull?.toEntity(),
    claimedModificationTime = claimedModificationTimeOrNull?.toInstant(),
    thumbnails = thumbnailsList.map { it.toEntity() },
    additionalClaimedMetadata = if (additionalClaimedMetadataList.isNotEmpty()) {
        additionalClaimedMetadataList.map { it.toEntity() }
    } else null,
    contentAuthor = contentAuthorOrNull?.toEntity(),
)
